File: NNPRO-employee-management-main/NNPRO-employee-management-main/implementation/backend/src/main/java/upce/backend/entity/Task.java
package upce.backend.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import upce.backend.enums.TaskDifficulty;
import upce.backend.enums.TaskStatus;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String title;

    private String description;

    @CreationTimestamp
    private LocalDateTime creationDate;

    private LocalDateTime dueDate;

    private LocalDateTime expectedFinishDate;

    private LocalDateTime finishDate;

    @Enumerated(EnumType.STRING)
    private TaskDifficulty difficulty;

    @Enumerated(EnumType.STRING)
    private TaskStatus status;

    @ManyToMany(mappedBy = "tasks")
    private List<AppUser> appUsers = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL,orphanRemoval = true)
    private List<Attendance> attendances = new ArrayList<>();

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PerformanceReport> performances = new ArrayList<>();

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Comment> comments = new ArrayList<>();

    @PreRemove
    private void removeTaskFromUsers() {
        for (AppUser user : appUsers) {
            user.getTasks().remove(this);
        }
    }
}

File: NNPRO-employee-management-main/NNPRO-employee-management-main/implementation/frontend/src/types/task.ts
import {PaginatedResponse} from "./pagination.ts";
import {Member} from "./member.ts";

export interface Task {
    id: number;
    title: string;
    description: string;
    dueDate: string;
    status: Status;
    finishDate: string;
    difficulty: Status;
    projectId: number;
    assignedUsers: Member[];
}

export interface TaskResponse {
    success: boolean;
    task?: Task;
    message?: string;
}

export interface TasksResponse {
    success: boolean;
    tasks?: Task[];
    message?: string;
    pagination?: PaginatedResponse;
}

export interface TaskStatusesResponse {
    success: boolean;
    statuses?: Status[];
    message?: string;
}
export interface TaskDifficultyResponse {
    success: boolean;
    statuses?: Status[];
    message?: string;
}

export interface Status {
    name: string;
    label: string;
}
